pidtree_bcc/plugin.py

```diff
@@ -0,0 +1,48 @@
+import importlib
+
+class BasePlugin:
+    def __init__(self, args):
+        self.validate_args(args)
+
+    def process(self, event):
+        """ Process the `event` dict, add in additional metadata and
+        return a dict """
+        raise NotImplementedError("Required method `process` has not been implemented by {}" % self.__name__)
+
+    def validate_args(self, args):
+        """ Not required, override in inheriting class if you want to
+        use this """
+        pass
+
+
+class Plugins:
+    def __init__(self, plugin_dict, plugin_dir="pidtree_bcc.plugins"):
+        """ `plugin_dict` is a dict where the keys are plugin names and
+        the value for each key is another dict of kwargs. Each key must
+        match a `.py` file in the plugin directory and the kwargs can
+        be validated on initialization of the plugin """
+        self._plugins = []
+        for plugin_name, plugin_args in plugin_dict.items():
+            if not plugin_args.get("enabled", False):
+                next
+            plugin_classname = plugin_name.capitalize()
+            import_line = ".".join([plugin_dir, plugin_name])
```

Comment:
I'm mixed on dynamically importing all plugins; could be bad if someone pushes a bad plugin that breaks everything. We'd have no way to whitelist the list of approved plugins.

I'm fine with this as is, or creating a hardcoded set of plugins that needed to get listed out explicitly whenever a new plugin is added if you anticipate that the set of plugins will grow to a sizeable number

Reply:
Well, plugins get loaded according to the configuration file. If you want to avoid a plugin being loaded it is sufficient just not to declare it.